### PR TITLE
Migrate to static config

### LINKED ISSUE #1966: Automated web application configuration migration
The SecureDrop web application's configuration is read from `config.py`, which is generated from a template, `config.py.example`, by the `initialize_securedrop_app` [task](https://github.com/freedomofpress/securedrop/blob/892cf52299f17ed9c70df0bb0bec68137101bab2/install_files/ansible-base/roles/app/tasks/initialize_securedrop_app.yml#L29-L151) during the initial run of the Ansible playbooks.

During the initial development of the SecureDrop auto-upgrade strategy, we faced a dilemma in managing `config.py`. The dilemma is mostly captured in comments ([1](https://github.com/freedomofpress/securedrop/blob/892cf52299f17ed9c70df0bb0bec68137101bab2/install_files/ansible-base/roles/app/tasks/initialize_securedrop_app.yml#L29-L33), [2](https://github.com/freedomofpress/securedrop/blob/892cf52299f17ed9c70df0bb0bec68137101bab2/install_files/ansible-base/roles/app/tasks/initialize_securedrop_app.yml#L59-L65)) in the relevant Ansible role, but I will summarize it here for clarity.

At the time of development, the most straightforward way to provide an auto-upgrade capability for SecureDrop was to use Debian packages. The SecureDrop application code is provided by the `securedrop-app-code` package, and production instances check https://apt.freedom.press periodically for updates to this package. Files that contain instance-specific configuration cannot be included in the Debian package because they would be overwritten on update.

Additionally, the instance-specific configuration depends on configuration values that must be provided by the SecureDrop administrator during installation, so the task of generating and maintaining the web application configuration fell to the Ansible playbooks, which are configured with vars by the administrator during installation and run from the Admin Workstation.

Finally, there are some configuration values which automatically generated by the Ansible playbooks. For security, these values must be unique, random, and generated from a secure CSPRNG (e.g. `/dev/urandom`). These values [are](https://github.com/freedomofpress/securedrop/blob/892cf52299f17ed9c70df0bb0bec68137101bab2/install_files/ansible-base/roles/app/tasks/initialize_securedrop_app.yml#L71-L133):

* `source secret key`—the Flask `SECRET_KEY` for the Source Interface, used as an HMAC key to secure Flask's sessions from tampering.
* `journalist secret key`—same, but for the Journalist Interface.
* `scrypt id pepper`—random value used as an instance-wide salt (aka "pepper") for deriving Source's filesystem ID's from their codenames.
* `scrypt gpg pepper`—random value used as an instance-wide salt (aka "pepper") for deriving Source's reply keypair passphrases from their codenames.

Once set, it is **critical** that these values not be modified during the lifetime of the instance. To be specific:

* `{source,journalist} secret key` could be modified with only minor disruption to the operation of the instance. Changing these values would invalidate all current sessions, which means any currently logged in users would encounter errors and be required to log in again.
* `scrypt {id,gpg} pepper` **cannot be modified** without severe disruption to the operation of the instance. Since these values are required to derive source ID's and reply keypair passphrases from their codenames, modifying them would make it impossible for sources who had created their accounts prior to the modification to access their accounts.

The Ansible playbooks do not copy these values back to the Admin Workstation as part of the installation process, although they are included in backups made by the Ansible backup role. Because it is so important to avoid accidentally overwriting these automatically generated values, the Ansible playbooks currently take a [defensive approach](https://github.com/freedomofpress/securedrop/blob/892cf52299f17ed9c70df0bb0bec68137101bab2/install_files/ansible-base/roles/app/tasks/initialize_securedrop_app.yml#L29-L41) and avoid interacting with `config.py` *at all* if it is found to already exist on the system.

Automated migration of the SecureDrop web application configuration is prevented by the confluence of all of these factors. In order to make it possible, we need to decide:

1. How should configuration be canonically represented?
    * Configuration is not just the values of variables in `config.py`: it is also any files that may be referenced by `config.py`, e.g. the GPG key whose fingerprint is referenced in `JOURNALIST_KEY`.
    * We have discussed how using a Python file to store configuration is problematic because it can (and currently does) contain conditional logic, which makes it harder to reason about the configuration and makes it hard to separate concerns.
2. Where should the canonical representation of the configuration be stored (e.g. the Admin Workstation, the servers, both)? Any component of the system that can auto-migrate the configuration will need a copy of the canonical representation to handle all possible migrations.
3. What components of the system should be allowed to automatically migrate the configuration?
    * This is tricky. On one hand, it could be useful to allow the `securedrop-app-code` package to automatically migrate the web application configuration. On the other hand, this creates a potential need for conflict resolution between disparate representations of an instance's configuration.

Clearly, this is a complex issue and further discussion and design effort is warranted. I think it should be possible to break this issue up into discrete sub-issues and resolve it incrementally, but we will need to carefully devise a plan to do so.

### PR DESCRIPTION
**Pull request by heartsucker**
## Status

Ready for review

## Description of Changes

Fixes #1966

Changes proposed in this pull request:

- Define `config.json` configuration file and fields
- Create a migration tool to convert old `config.py` into new `config.json`
- Set defaults at runtime
- Populate missing config values via `ansible`

### New Installs
- Checkout this branch
- Build debs, provision staging VM
- Check that app works
- Back up the configs
- `dpkg -i` the app code deb (or just run `postinst`)
- Check that the it was idempotent and the configs weren't changed

### Old Installs (with ansible)

- Checkout the commit on `develop` that this branch
- Build debs, provision staging VM
- Check out the head of this branch
- Build debs, provision new VM
- Check that app works
- Check that config was migrated and contains the same values as `config.py`
- Back up the new configs to `/tmp` or `~`
- `dpkg -i` the app code deb (or just run `postinst`)
- Check that the it was idempotent and the configs weren't changed

### Old Installs (without ansible)

- Checkout the commit on `develop` that this branch
- Build debs, provision staging VM
- Check out the head of this branch
- Build debs, `scp` to VM
- Install debs manually (to simulate `cron-apt`)
- Check that app works
- Check that config was migrated and contains the same values as `config.py`
- Back up the new configs to `/tmp` or `~`
- `dpkg -i` the app code deb (or just run `postinst`)
- Check that the it was idempotent and the configs weren't changed

### Old Installs Backup/Restore

- Checkout the commit on `develop` that this branch
- Build debs, provision staging VM
- Run backup script
- Check out the head of this branch
- Build debs, provision new VM
- `rm /etc/securedrop/*.json`
- Run restore script
- Check that config was migrated and contains the same values as `config.py`

## Deployment

We need to be able to support the following cases:

1. `config.py` is not present
2. `config.py` present, `config.py` has partial configuration
3. `config.json` present, `config.json` has partial configuration

### If you made changes to the server application code:

- [x] Linting (`make ci-lint`) and tests (`make -C securedrop test`) pass in the development container

### If you made changes to the system configuration:

- [x] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass

### If you made non-trivial code changes:

- [x] I have written a test plan and validated it for this PR

### CODE REVIEW

**Comment by eloquence** [comment]
For the 11/28-12/12, we will aim to do a first time-boxed review; not planning to merge yet.

**Comment by kushaldas** [comment]
I was planning to make the `SDConfig` class into a proper strictly defined class (this is just a comment).

File: molecule/testinfra/staging/app-code/test_securedrop_app_code.py [hunk 1]
```diff
@@ -34,18 +35,6 @@ def test_securedrop_application_apt_dependencies(Package, package):
...
-def test_securedrop_application_test_locale(File, Sudo):
-    """
-    Ensure SecureDrop DEFAULT_LOCALE is present.
-    """
-    securedrop_config = File("{}/config.py".format(
-        securedrop_test_vars.securedrop_code))
-    with Sudo():
-        assert securedrop_config.is_file
-        assert securedrop_config.contains("^DEFAULT_LOCALE")
-        assert securedrop_config.content.count("DEFAULT_LOCALE") == 1
```
[the remark this answers is not in the document]
Comment (heartsucker) [reply molecule/testinfra/staging/app-code/test_securedrop_app_code.py]: I don't think we need these tests because the new `Config` class will gracefully handle the missing value and set it to `en_US`.

**Comment by heartsucker** [comment]
Before this is merged, an open question is: should `config.json` be split into `source-config.json` and `journalist-config.json`? I swear there was a reason I didn't do this at first, but I can't for the life of me remember what that might have been. Probably that this was just simpler? The advantage of this would be in the future we could have stricter permissions / AppArmor rules to prevent the source app from reading secret data from the journalist config.

File: install_files/ansible-base/roles/backup/files/backup.py [hunk 2]
```diff
@@ -18,13 +18,15 @@ def main():
...
-    sd_config = os.path.join(sd_code, "config.py")
+    sd_config_py = os.path.join(sd_code, "config.py")
+    sd_config = os.path.join("/etc/securedrop/config.json")
     sd_custom_logo = os.path.join(sd_code, "static/i/logo.png")
 
     tor_hidden_services = "/var/lib/tor/services"
     torrc = "/etc/tor/torrc"
 
     with tarfile.open(backup_filename, 'w:gz') as backup:
+        backup.add(sd_config_py)
```
[the remark this answers is not in the document]
Comment (heartsucker) [reply install_files/ansible-base/roles/backup/files/backup.py]: see the logic in `restore.py` to see if this does the trick.

**Comment by kushaldas** [comment]
>Before this is merged, an open question is: should config.json be split into source-config.json and journalist-config.json? I swear there was a reason I didn't do this at first, but I can't for the life of me remember what that might have been. Probably that this was just simpler? The advantage of this would be in the future we could have stricter permissions / AppArmor rules to prevent the source app from reading secret data from the journalist config.

As we discussed last night, please go ahead and split them across into two files. This PR also needs `Test steps`.

**Comment by heartsucker** [comment]
@kushaldas @redshiftzero I've made a few more changes. Basically all all the config generation logic has been moved into `populate_config.py` and doesn't exist in ansible (other than as a shell-out to it). There were some tests added for a couple of common cases for the state of the config files, but there are tons of edge cases and I'm not sure how thorough we want to be about the entire space of possibilities for this.

**Comment by heartsucker** [comment]
Ah, also the configs were split into a source and journalist version, both as python classes and different JSON files. I also stripped out all global references to the config and everything is injected (for tests and the application).

File: install_files/ansible-base/roles/backup/files/backup.py [hunk 3]
```diff
@@ -18,13 +18,15 @@ def main():
     sd_data = '/var/lib/securedrop'
 
     sd_code = '/var/www/securedrop'
-    sd_config = os.path.join(sd_code, "config.py")
+    sd_config_py = os.path.join(sd_code, "config.py")
+    sd_config = os.path.join("/etc/securedrop/config.json")
```
Comment (kushaldas) [inline install_files/ansible-base/roles/backup/files/backup.py]: These config json file paths should be updated with the new 2 files.

File: molecule/builder/tests/test_securedrop_deb_package.py [hunk 4]
```diff
@@ -151,15 +151,16 @@ def test_deb_package_control_fields_homepage(File, Command, deb):
 def test_deb_package_contains_no_config_file(File, Command, deb):
     """
     Ensures the `securedrop-app-code` package does not ship a `config.py`
-    file. Doing so would clobber the site-specific changes made via Ansible.
+    file or a `config.json`. Doing so would clobber the site-specific changes
```
Comment (kushaldas) [inline molecule/builder/tests/test_securedrop_deb_package.py]: Update comment to have two different file names.

File: molecule/testinfra/staging/app-code/test_securedrop_app_code.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kushaldas) [reply molecule/testinfra/staging/app-code/test_securedrop_app_code.py]: >I don't think we need these tests because the new Config class will gracefully handle the missing value and set it to en_US.

Correct, but, please make sure that is there is a test which checks for this default language value.

File: molecule/testinfra/staging/app-code/test_securedrop_app_code.py [hunk 5]
```diff
@@ -64,17 +53,22 @@ def test_securedrop_application_test_journalist_key(File, Sudo):
 
     # Let's make sure the corresponding fingerprint is specified
     # in the SecureDrop app configuration.
-    securedrop_config = File("{}/config.py".format(
-        securedrop_test_vars.securedrop_code))
+    securedrop_config = File("/etc/securedrop/config.json")
```
Comment (kushaldas) [inline molecule/testinfra/staging/app-code/test_securedrop_app_code.py]: This test will also need update.

File: securedrop/i18n.py [hunk 6]
```diff
@@ -39,7 +39,7 @@ def setup_app(config, app):
     global LOCALES
     global babel
 
-    translation_dirs = getattr(config, 'TRANSLATION_DIRS', None)
+    translation_dirs = config.TRANSLATION_DIRS
```
Comment (kushaldas) [inline securedrop/i18n.py]: @heartsucker I can not explain how happy I am to see these `*attr` calls being removed :heart: .

File: securedrop/sdconfig.py [hunk 7]
```diff
@@ -11,154 +14,216 @@
...
+    WTF_CSRF_ENABLED = True
+
+    def __init__(self, secret_key, cookie_name, env):
+        # type: (str, str, str) -> None
+        self.SECRET_KEY = secret_key
+        self.SESSION_COOKIE_NAME = cookie_name
+
+        if env == 'prod':
+            # This is recommended for performance, and also resolves #369
+            self.USE_X_SENDFILE = True  # type: ignore
```
Comment (kushaldas) [inline securedrop/sdconfig.py]: Can we please have all the possible attributes of the class with some default values? Let us say, locked down production values, and only enable/set in the other cases. Will really help to read/type annotations.

File: securedrop/sdconfig.py [hunk 8]
```diff
@@ -11,154 +14,216 @@
...
-        except AttributeError:
-            pass
+        with open(config_file) as f:
+            self._json_config = json.loads(f.read())
 
-        try:
-            self.DATABASE_USERNAME = _config.DATABASE_USERNAME  # type: ignore
-        except AttributeError:
-            pass
+        self.SECUREDROP_DATA_ROOT = '/var/lib/securedrop/'
```
Comment (kushaldas) [inline securedrop/sdconfig.py]: Please add the type annotations.

File: securedrop/sdconfig.py (hunk 7 above)
Comment (heartsucker) [reply securedrop/sdconfig.py]: Flask handles the default on its own, and I don't think we should set them ourselves.

File: securedrop/sdconfig.py (hunk 8 above)
Comment (heartsucker) [reply securedrop/sdconfig.py]: Can you check again that sufficient type annotations have been added?

File: securedrop/sdconfig.py (hunk 7 above)
Comment (kushaldas) [reply securedrop/sdconfig.py]: @heartsucker I am asking defaults to enable/use `mypy` in a better way.

File: install_files/securedrop-app-code/DEBIAN/postinst [hunk 9]
```diff
@@ -85,6 +85,9 @@ case "$1" in
     done
 
     chown -R www-data:www-data /var/lib/securedrop /var/www/securedrop
+    
+    mkdir -p /etc/securedrop
+    chdmod 0750 /etc/securedrop
```
Comment (emkll) [inline install_files/securedrop-app-code/DEBIAN/postinst]: Is this meant to be `chmod`?

**Comment by eloquence** [comment]
Recapping our discussion from sprint planning today:

We want to minimize the risks associated with this big (and important!) change by ensuring we express through automated tests the potential states that the configuration file can be in. The number of states should be manageable given that the file is not manually edited: when have the playbooks been run, and what features have been explicitly enabled? We must assume that there are long-running instances that have not run playbooks since pre-0.4. 

@heartsucker -- on the FPF side, we've committed to up to 8 person hours to help get this closer to the finish line, but we should make sure we're on the same page vis-a-vis the testing strategy. Let's discuss w/ @redshiftzero on Gitter, or during standup, next time you're around.

**Comment by heartsucker** [comment]
As a note to reviewers, the test `test_populate_config.py` is the primary place one would want to look to check for what states of config files we are currently operating under. Additional tests would probably be added here.

**Comment by eloquence** [comment]
Hi @heartsucker, this has been sitting for a long time and I think it's time to acknowledge that we're probably not going to land a version of this change, which would require significant PR surgery at this point, and which is quite high-risk in terms of potential for breaking existing configs.

Instead our bias is to prioritize moving as many settings as reasonably possible to the database (where we can use alembic migrations for change management) and the web-based admin UI.

I'm going to leave this open a few more days in case other folks want to make the case for revisiting this PR, and then close.

**Comment by eloquence** [comment]
Closing PR per the above comment; let's keep the `static-config` branch around for now in case we want to revisit at a later date.
